### ISSUE TITLE
Custom operators registered via decorator slower than ops registered via `torch.Library.{define, impl}`

### ISSUE THREAD

**Issue by bnellnm** [pytorch/pytorch#139500]
### 🐛 Describe the bug

Custom operators registered via `torch.library.custom_op` seem to be much slower than ops registered via `torch.Library.define` + `torch.Library.impl`

```
import torch

@torch.library.custom_op("foo::bar", mutates_args=())
def bar(a: torch.Tensor,
        b: torch.Tensor,
        c: torch.Tensor,
        d: torch.Tensor,
        e: torch.Tensor,
        f: torch.Tensor,
        g: torch.Tensor,
        h: torch.Tensor,
        i: torch.Tensor,
        j: torch.Tensor,
        k: torch.Tensor,
        l: torch.Tensor,
        m: torch.Tensor,
        n: torch.Tensor) -> torch.Tensor:
    return b.clone()

@torch.library.custom_op("foo::baz", mutates_args=(["a"]))
def baz(a: torch.Tensor,
        b: torch.Tensor,
        c: torch.Tensor,
        d: torch.Tensor,
        e: torch.Tensor,
        f: torch.Tensor,
        g: torch.Tensor,
        h: torch.Tensor,
        i: torch.Tensor,
        j: torch.Tensor,
        k: torch.Tensor,
        l: torch.Tensor,
        m: torch.Tensor,
        n: torch.Tensor) -> torch.Tensor:
    return b.clone()

def barbaz(a: torch.Tensor,
        b: torch.Tensor,
        c: torch.Tensor,
        d: torch.Tensor,
        e: torch.Tensor,
        f: torch.Tensor,
        g: torch.Tensor,
        h: torch.Tensor,
        i: torch.Tensor,
        j: torch.Tensor,
        k: torch.Tensor,
        l: torch.Tensor,
        m: torch.Tensor,
        n: torch.Tensor) -> torch.Tensor:
    return b.clone()

foo_lib = torch.library.Library("foo", "FRAGMENT")

def direct_register_custom_op(
    op_name,
    op_func,
    mutates_args
):
    schema_str = torch.library.infer_schema(op_func, mutates_args=mutates_args)
    foo_lib.define(op_name + schema_str)
    foo_lib.impl(op_name, op_func, "CUDA")

direct_register_custom_op("foo::bar_op", barbaz, mutates_args=())
direct_register_custom_op("foo::baz_op", barbaz, mutates_args=(["a"]))

a = torch.rand([128,128], device="cuda")
b = torch.rand([128,128], device="cuda")
c = torch.rand([128,128], device="cuda")
d = torch.rand([128,128], device="cuda")
e = torch.rand([128,128], device="cuda")
f = torch.rand([128,128], device="cuda")
g = torch.rand([128,128], device="cuda")
h = torch.rand([128,128], device="cuda")
i = torch.rand([128,128], device="cuda")
j = torch.rand([128,128], device="cuda")
k = torch.rand([128,128], device="cuda")
l = torch.rand([128,128], device="cuda")
m = torch.rand([128,128], device="cuda")
n = torch.rand([128,128], device="cuda")


def test():
    from triton.testing import do_bench

    iter = 1000

    def mutate():
        for z in range(iter):
            o = torch.ops.foo.baz(a, b, c, d, e, f, g, h, i, j, k, l, m, n)

    def no_mutate():
        for z in range(iter):
            o = torch.ops.foo.bar(a, b, c, d, e, f, g, h, i, j, k, l, m, n)

    def direct_mutate():
        for z in range(iter):
            o = torch.ops.foo.baz_op(a, b, c, d, e, f, g, h, i, j, k, l, m, n)

    def direct_no_mutate():
        for z in range(iter):
            o = torch.ops.foo.bar_op(a, b, c, d, e, f, g, h, i, j, k, l, m, n)

    mutate_time = do_bench(mutate)
    no_mutate_time = do_bench(no_mutate)
    direct_mutate_time = do_bench(direct_mutate)
    direct_no_mutate_time = do_bench(direct_no_mutate)

    print(f"mutate = {mutate_time}")
    print(f"no_mutate = {no_mutate_time}")
    print(f"direct_mutate = {direct_mutate_time}")
    print(f"direct_no_mutate = {direct_no_mutate_time}")

test()
```

Running the script gives me the following results:
```
mutate = 90.21110534667969
no_mutate = 25.86481285095215
direct_mutate = 6.907863140106201
direct_no_mutate = 6.97034215927124
```

### Versions

Collecting environment information...
PyTorch version: 2.5.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.30.5
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.5.0-35-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.5.82
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H100 80GB HBM3
GPU 1: NVIDIA H100 80GB HBM3
GPU 2: NVIDIA H100 80GB HBM3
GPU 3: NVIDIA H100 80GB HBM3
GPU 4: NVIDIA H100 80GB HBM3
GPU 5: NVIDIA H100 80GB HBM3
GPU 6: NVIDIA H100 80GB HBM3
GPU 7: NVIDIA H100 80GB HBM3

Nvidia driver version: 555.42.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.2.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             128
On-line CPU(s) list:                0-127
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Platinum 8462Y+
CPU family:                         6
Model:                              143
Thread(s) per core:                 2
Core(s) per socket:                 32
Socket(s):                          2
Stepping:                           8
CPU max MHz:                        4100.0000
CPU min MHz:                        800.0000
BogoMIPS:                           5600.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hfi vnmi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr ibt amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          3 MiB (64 instances)
L1i cache:                          2 MiB (64 instances)
L2 cache:                           128 MiB (64 instances)
L3 cache:                           120 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-31,64-95
NUMA node1 CPU(s):                  32-63,96-127
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] flashinfer==0.1.2+cu121torch2.4
[pip3] mypy==1.11.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] onnx==1.14.1
[pip3] onnxruntime==1.18.1
[pip3] torch==2.5.0
[pip3] torchvision==0.20.0
[pip3] triton==3.1.0
[conda] Could not collect                                       

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @chauhang @penguinwu @bdhirsh @yf225

**Comment by zou3519** [comment]
By the way, you should be able to get a little more perf by using torch.ops.foo.bar.default (with the extra default at the end) (cc @youkaichao too)

**Comment by zou3519** [comment]
@bnellnm the overhead tends to increase depending on the number of Tensor inputs. Is 14 tensor inputs (what is in the repro) a reasonable workload for the custom ops you're working with?

**Comment by bnellnm** [comment]
> [@bnellnm](https://github.com/bnellnm) the overhead tends to increase depending on the number of Tensor inputs. Is 14 tensor inputs (what is in the repro) a reasonable workload for the custom ops you're working with?

The 14 input one was an extreme case (paged attention IIRC?).  Most ops have fewer inputs although ~8-10 is not unusual.

**Comment by zou3519** [comment]
@stadlmax the autograd registration is there for safety. We can add a way to remove it, or investigate a way to disable it if we're under inference_mode (where gradient computation isn't needed)

**Comment by xmfan** [comment]
from triage: @azahed98 is exploring ways to make the AOTAutograd wrapper faster, the custom ops runtime wrapper here may benefit from that work. Arsh, could you circle back with the findings afterwards? thanks!

**Comment by ezyang** [comment]
Making AOTAutograd wrappers faster won't really help here though...

**Comment by zou3519** [comment]
torch.library.custom_op has additional overhead from the following:
1) a default autograd not implemented registration
2) extra O(#input+#output) tensor checking.

For (1) the solution is to be able to register a custom operator for inference only (e.g. disable all the autograd not implemented checks)
For (2), maybe a flag to ignore that too.

**Comment by zou3519** [comment]
NB: I think Arsh is working on AOTAutograd overhead, not custom operator overhead, so I'm going to unassign for now

**Comment by riccardofelluga** [comment]
Hi! I'd like to take a look :) 

cc. @kshitij12345
